src/mscorlib/src/System/String.Manipulation.cs

```diff
@@ -951,7 +951,173 @@ public String Remove(int startIndex, int count)
             Contract.EndContractBlock();
 
             return Substring(0, startIndex);
-        }   
+        }
+
+        public string Replace(string oldValue, string newValue, bool ignoreCase, CultureInfo culture)
+        {
+            Contract.Ensures(Contract.Result<String>() != null);
+            Contract.EndContractBlock();
+
+            return ReplaceCore(oldValue, newValue, culture, ignoreCase ? CompareOptions.IgnoreCase : CompareOptions.None);
+        }
+
+        public string Replace(string oldValue, string newValue, StringComparison comparison)
+        {
+            Contract.Ensures(Contract.Result<String>() != null);
+            Contract.EndContractBlock();
+
+            switch (comparison)
+            {
+                case StringComparison.CurrentCulture:
+                    return ReplaceCore(oldValue, newValue, CultureInfo.CurrentCulture, CompareOptions.None);
+
+                case StringComparison.CurrentCultureIgnoreCase:
+                    return ReplaceCore(oldValue, newValue, CultureInfo.CurrentCulture, CompareOptions.IgnoreCase);
+
+                case StringComparison.InvariantCulture:
+                    return ReplaceCore(oldValue, newValue, CultureInfo.InvariantCulture, CompareOptions.None);
+
+                case StringComparison.InvariantCultureIgnoreCase:
+                    return ReplaceCore(oldValue, newValue, CultureInfo.InvariantCulture, CompareOptions.IgnoreCase);
+
+                case StringComparison.Ordinal:
+                    return ReplaceCore(oldValue, newValue, CultureInfo.InvariantCulture, CompareOptions.Ordinal);
+
+                case StringComparison.OrdinalIgnoreCase:
+                    return ReplaceCore(oldValue, newValue, CultureInfo.InvariantCulture, CompareOptions.OrdinalIgnoreCase);
+
+                default:
+                    throw new ArgumentException(Environment.GetResourceString("NotSupported_StringComparison"), nameof(comparison));
+            }
+        }
+
+        private unsafe String ReplaceCore(String oldValue, String newValue, CultureInfo culture, CompareOptions compareOptions)
+        {
+            if (oldValue == null)
+                throw new ArgumentNullException(nameof(oldValue));
+
+            if (newValue == null)
+                newValue = string.Empty;
+
+            CultureInfo referenceCulture = culture ?? CultureInfo.CurrentCulture;
+
+            List<KeyValuePair<int, int>> replacementMarkers = new List<KeyValuePair<int, int>>(20);
```

Comment:
>List<KeyValuePair<int, int>> replacementMarkers = new List<KeyValuePair<int, int>>(20); [](start = 12, length = 87)

I think it would be simpler if you implement this with StringBuilder. here is some pseudo code 

```C#

StringBuilder sb = new StringBuilder();
int searchIndex = 0;
int index;

do 
{
    index  = IndexOfCore(... ref matchLength);
    if (index >= 0)
    {
        sb.Append(this, searchIndex, index);
        sb.Append(newValue);
        searchIndex += index + matchLength;
    }
    else
    {
        sb.Append(this, searchIndex, this.Length - searchIndex);
    }
} while (index >= 0);
```

Note that, we need to optimize the case if we didn't find any match we should return the original string